echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/notification/CDEventsNotificationAgent.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.notification;
+
+import com.netflix.spinnaker.echo.api.events.Event;
+import com.netflix.spinnaker.echo.cdevents.CDEventsBuilderService;
+import com.netflix.spinnaker.echo.cdevents.CDEventsSenderService;
+import dev.cdevents.exception.CDEventsException;
+import io.cloudevents.CloudEvent;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Map;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Service;
+import retrofit.mime.TypedByteArray;
+
+@Slf4j
+@ConditionalOnProperty("cdevents.enabled")
+@Service
+public class CDEventsNotificationAgent extends AbstractEventNotificationAgent {
+  @Autowired CDEventsBuilderService cdEventsBuilderService;
+  @Autowired CDEventsSenderService cdEventsSenderService;
+
+  @Override
+  public String getNotificationType() {
+    return "cdevents";
+  }
+
+  @Override
+  public void sendNotifications(
+      Map<String, Object> preference,
+      String application,
+      Event event,
+      Map<String, String> config,
+      String status) {
+    log.info("Sending CDEvents notification..");
+
+    String executionId =
+        Optional.ofNullable(event.content)
+            .map(e -> (Map) e.get("execution"))
+            .map(e -> (String) e.get("id"))
+            .orElse(null);
+    String cdEventsType =
+        Optional.ofNullable(preference).map(p -> (String) p.get("cdEventsType")).orElse(null);
+    String eventsBrokerUrl =
+        Optional.ofNullable(preference).map(p -> (String) p.get("address")).orElse(null);
+
+    try {
+      CloudEvent ceToSend =
+          cdEventsBuilderService.createCDEvent(
+              preference, application, event, config, status, getSpinnakerUrl());
+      log.info(
+          "Sending CDEvent {} notification to events broker url {}", cdEventsType, eventsBrokerUrl);
+      HttpURLConnection response =
+          cdEventsSenderService.sendCDEvent(ceToSend, new URL(eventsBrokerUrl));
+      if (response != null) {
+        log.info(
+            "Received response from events broker : {} {} for execution id {}. {}",
+            response.getResponseCode(),
+            response.getResponseMessage(),
+            executionId,
+            new String(((TypedByteArray) response.getContent()).getBytes()));
+      }
+
+    } catch (Exception e) {
+      log.error("Exception occurred while sending CDEvent {}", e);
+      throw new CDEventsException("Exception occurred while sending CDEvent", e);
```

Comment:
As a rule, we should be throwing spinnaker exceptions not custom exceptions for these.

Reply:
One thing to call out here, is unless you are modifying the retrofit exception handler, these custom exceptions may not be handled properly (I dont think). One thing we could consider doing is having some base spinnaker exception that custom exceptions could inherit from which could make throwing custom exceptions a little more flexible

Reply:
`CDEventsException` is not needed now as using the retrofit request/response to send CDEvents notification